File: Models/SrtSubtitleLine.vb
Namespace Models
    ''' <summary>
    ''' Dòng phụ đề định dạng SRT
    ''' Format:
    '''   Index (số thứ tự)
    '''   HH:MM:SS,fff --> HH:MM:SS,fff
    '''   Text
    ''' </summary>
    Public Class SrtSubtitleLine
        Inherits SubtitleLine

        Public Property Index As Integer
        Public Property Text As String

        ''' <summary>
        ''' Constructor rỗng
        ''' </summary>
        Public Sub New()
        End Sub

        ''' <summary>
        ''' Constructor từ các thành phần
        ''' </summary>
        Public Sub New(index As Integer, startTime As TimeSpan, endTime As TimeSpan, text As String)
            Me.Index = index
            Me.StartTime = startTime
            Me.EndTime = endTime
            Me.Text = text
        End Sub

        ''' <summary>
        ''' Rebuild lại dòng SRT với time mới
        ''' </summary>
        Public Overrides Function RebuildLine() As String
            Dim timeLine = String.Format("{0} --> {1}", FormatSrtTime(StartTime), FormatSrtTime(EndTime))
            Return String.Format("{0}{1}{2}{1}{3}", Index, Environment.NewLine, timeLine, Text)
        End Function

        ''' <summary>
        ''' Clone dòng SRT
        ''' </summary>
        Public Overrides Function Clone() As SubtitleLine
            Dim newLine = New SrtSubtitleLine()
            newLine.Index = Me.Index
            newLine.StartTime = Me.StartTime
            newLine.EndTime = Me.EndTime
            newLine.Text = Me.Text
            newLine.OriginalText = Me.OriginalText
            newLine.Content = Me.Content
            Return newLine
        End Function
    End Class
End Namespace

File: Models/SrtSubtitleLine.cs
using System;
using Subtitle_draft_GMTPC.Models;

namespace Subtitle_draft_GMTPC.Models
{
    /// <summary>
    /// Dòng phụ đề định dạng SRT
    /// Format:
    ///   Index (số thứ tự)
    ///   HH:MM:SS,fff --> HH:MM:SS,fff
    ///   Text
    /// </summary>
    public class SrtSubtitleLine : SubtitleLine
    {
        public int Index { get; set; }
        public string Text { get; set; }

        /// <summary>
        /// Constructor rỗng
        /// </summary>
        public SrtSubtitleLine()
        {
        }

        /// <summary>
        /// Constructor từ các thành phần
        /// </summary>
        public SrtSubtitleLine(int index, TimeSpan startTime, TimeSpan endTime, string text)
        {
            Index = index;
            StartTime = startTime;
            EndTime = endTime;
            Text = text;
        }

        /// <summary>
        /// Rebuild lại dòng SRT với time mới
        /// </summary>
        public override string RebuildLine()
        {
            string timeLine = $"{FormatSrtTime(StartTime)} --> {FormatSrtTime(EndTime)}";
            return $"{Index}{Environment.NewLine}{timeLine}{Environment.NewLine}{Text}";
        }

        /// <summary>
        /// Clone dòng SRT
        /// </summary>
        public override SubtitleLine Clone()
        {
            return new SrtSubtitleLine()
            {
                Index = this.Index,
                StartTime = this.StartTime,
                EndTime = this.EndTime,
                Text = this.Text,
                OriginalText = this.OriginalText,
                Content = this.Content
            };
        }
    }
}
